src/gui.rs

```diff
@@ -394,195 +375,26 @@ impl Sam {
-                }
-
-                ui.horizontal(|ui| {
-                    ui.label(*title);
-                    ui.with_layout(Layout::right_to_left(Align::Center), |ui| {
-                        if ui.add_enabled(flash.is_some(), Button::new("🖴  Flash")).clicked() {
-                            self.open_archive_log(flash.unwrap());
-                        }
-
-                        if ui.add_enabled(telem.is_some(), Button::new("📡 Telemetry")).clicked() {
-                            self.open_archive_log(telem.unwrap());
-                        }
-                    });
-                });
-            }
-
-            ui.add_space(10.0);
-            ui.horizontal(|ui| {
-                ui.add_visible_ui(self.archive_progress.is_some(), |ui| {
-                    let (done, total) = self.archive_progress.unwrap_or((0, 0));
-                    let f = (total > 0).then(|| done as f32 / total as f32).unwrap_or(0.0);
-                    let text = format!("{:.2}MiB / {:.2}MiB", done as f32 / (1024.0*1024.0), total as f32 / (1024.0*1024.0));
-                    ui.add_sized([ui.available_width(), 20.0], ProgressBar::new(f).text(text));
-                });
-            });
-            ui.add_space(10.0);
+        open_archive_window(ctx, &mut archive_open, self);
 
-            ui.checkbox(&mut self.replay_logs, "Replay logs");
-        });
         self.archive_window_open = archive_open;
 
         if self.data_source.simulation_settings().is_some() {
-            let old_settings = self.data_source.simulation_settings().unwrap().clone();
-
-            egui::SidePanel::left("sim").min_width(300.0).max_width(500.0).resizable(true).show(ctx, |ui| {
-                ui.set_enabled(!self.archive_window_open);
-                ui.heading("Simulation");
-                ui.add_space(20.0);
-
-                CollapsingHeader::new("Simulation Parameters").default_open(true).show(ui, |ui| {
-                    let settings = self.data_source.simulation_settings().unwrap();
-                    settings.ui(ui)
-                });
-
-                CollapsingHeader::new("(Simulated) FC Settings").default_open(false).show(ui, |ui| {
-                    let settings = self.data_source.simulation_settings().unwrap();
-                    settings.fc_settings.ui(ui, None)
-                });
-
-                ui.add_space(20.0);
-
-                let changed = *self.data_source.simulation_settings().unwrap() != old_settings;
-                let released = false;
-                ui.horizontal(|ui| {
-                    if ui.button("Reset").clicked() {
-                        *(self.data_source.simulation_settings().unwrap()) = SimulationSettings::default();
-                    }
-
-                    if ui.button("↻  Rerun").clicked() || (changed && released) {
-                        self.data_source.reset();
-                    }
-                });
-            });
+            create_simulation_panel(self, ctx);
```

Comment:
This might be a place where it would be possible to get rid of passing self, by just passing a boolean for enabling and the data source. I also was never a huge fan of how we check for a simulation data source here. Ideally we should try downcasting the boxed trait inside `self.data_source` to the actual `SimulationDataSource` type, and then passing that. But again, that's not necessarily relevant to the cleanup.